### ISSUE TITLE
Playlist in sequential order not played sequentially anymore

### ISSUE THREAD

**Issue by rquara** [AzuraCast/AzuraCast#4170]
**Using Docker installation method?**
Yes

**AzuraCast version:**
AzuraCast • Rolling Release #2f0b109 (2021-05-12 14:31) • Docker • PHP 8.0

**Host Operating System:**
Ubuntu 20.04

**Describe the bug:**
The files in a sequential playlist are not played in a sequential order anymore after the latest azuracast update was performed.
I already cleaned the docker cache but it didn't change anything.

**Steps to reproduce:**
1. Rename the files as follow: 01.mp3 02.mp3 03.mp3 04.mp3
2. upload the files in a folder named "sequentialplaylist"
3. create a sequential playlist called "sequentialplaylist" (see screenshots below for the settings used)
4. assign to the "sequentialplaylist" folder the "sequentialplaylist" playlist
5. wait until the playlist is played or check the "reorder" screen 

**Expected behavior:**
As it was the case in the pastm I expect the playlist to be played in the following order:
01.mp3
02.mp3
03.mp3
04.mp3

It doesn't happen and the order is different. I can't clearly identify if there is any ratio in the order but it isn't the expected one.
The order is different from what expected also in the "reorder" screen and even before the playlist going live.

**Relevant logs:**
NA

**Screenshots:**
![image](https://user-images.githubusercontent.com/64473163/118297468-7b577280-b4de-11eb-97f2-16dbfdcdb7dc.png)
![image](https://user-images.githubusercontent.com/64473163/118297493-827e8080-b4de-11eb-8d8e-96f147c3ffdd.png)
![image](https://user-images.githubusercontent.com/64473163/118297514-8a3e2500-b4de-11eb-9699-59af07a31db7.png)

**Device(s):**
The mp3 upload is performed (as always done) using Chrome 90.0.4430.212 (64 bit) on Windows 10

**Additional context:**
NA

**Comment by sidney65** [comment]
Hi , 

have a same problem already reported here : [](https://github.com/AzuraCast/AzuraCast/issues/3201)

I have try to program a simple play list on demo site ; the result is the same : only the first track are played ; the others are ignored  .
Sequential mode is the simple but essential mode for radio station.
Hope you find a solution soon .
thanks to the azuracast team for is work !

**Comment by rquara** [comment]
I just updated to Rolling Release #a665c80 (2021-06-02 1:34) • Docker • PHP 8.0
I'm still experiencing the same problem.

Did anybody find a workaround beside of manually re-ordening the playlist? 

Thanks

**Comment by SlvrEagle23** [comment]
Frustratingly, I can't reproduce this issue locally...if I upload a playlist, set it as sequential and then follow the queue to see how it follows the playback order, it always plays through the entire playlist, in order, before resetting and returning to the start of the sequence, exactly as expected.

**Comment by rquara** [comment]
> Frustratingly, I can't reproduce this issue locally...if I upload a playlist, set it as sequential and then follow the queue to see how it follows the playback order, it always plays through the entire playlist, in order, before resetting and returning to the start of the sequence, exactly as expected.

is there any log or anything else I can share with you? I can also give you access to our test radio (running on the same instance but not advertised)

**Comment by Marcioghiraldelli** [comment]
if you create one or two lists, maybe even 3 lists it can even follow normally, but if you have several lists, 10, 15, 20 or even 30 lists it does that, it doesn't work correctly, mixes files or doesn't work sequentially , this only in auto dj managed by azuracast, if you put auto dj without azura gerencimaneto it follows normal.

**Comment by rquara** [comment]
> if you create one or two lists, maybe even 3 lists it can even follow normally, but if you have several lists, 10, 15, 20 or even 30 lists it does that, it doesn't work correctly, mixes files or doesn't work sequentially , this only in auto dj managed by azuracast, if you put auto dj without azura gerencimaneto it follows normal.


I'm not sure if it is the case.
I tested the scenario on a test radio and I had just 1 active playlist (and 6 disabled playlists). The order was still not correct.

**Comment by rquara** [comment]
I just upgraded to Rolling Release #8a6625f (2021-06-21 8:15) • Docker • PHP 8.0 and the playlists order is still not correct.
I also flushed the docker cache and restarted the server after the upgrade

**Comment by Marcioghiraldelli** [comment]
In azuracast managed mode, sequential mode does not work well. You need to turn off this mode and use liquidsoap managed mode.

**Comment by hottweelz** [comment]
I've tried everything with azuracast, docker, ansible, every playlist option in the world.

The ONLY way I can get "Sequential" mp3s to play is via "General Playlist" none of the other options work (per x minutes, per x songs, etc)

new VMs, upgrades, downgrades, all of it... it's broken.

Can someone help give me a sample of four mp3s in sequential, once every X minutes referencing this: "You need to turn off this mode and use liquidsoap managed mode."

what would be the liquidsoap lines to add?  @Marcioghiraldelli @SlvrEagle23 ?

I'm desperate, frustrated and need help.  thanks so much in advance!   -james

**Comment by Marcioghiraldelli** [comment]
@hottweelz 
The only way I was able to do this is by turning off azuracast's managed mode. Once this is done, you mark sequentially, define the way you want to enter the intervals, you need to see also how azuracast is setting up your playlist in a playlist, then option more and then playlist. if it's not in sequence you should arrange in reorder.
Remember that this mode I mentioned, to apply changes you need to restart the radio.

**Comment by hottweelz** [comment]
@Marcioghiraldelli I appreciate it.  What I'm having trouble understanding is:  (and this is for advertising sequences, we switched providers where we play an intro.mp3, a filler_sweep.mp3, and then an outro.mp3 in that order)

Am I turning on Managed Mode on the individual playlist or the Advanced Manual DJ Mode under AutoDJ in the Profile Settings?

Also, we have to make a custom list like below?  What would it look like for say every 30 mins?

playlist_ads_new = playlist.once(id="playlist_ads_new",reload_mode="watch","/var/azuracast/stations/xtreme_free4all/playlists/playlist_ads_new.m3u")
playlist_ads_new = audio_to_stereo(id="stereo_playlist_ads_new", playlist_ads_new)
playlist_ads_new = cue_cut(id="cue_playlist_ads_new", playlist_ads_new)
ignore(playlist_ads_new)

**Comment by Marcioghiraldelli** [comment]
if you want to play 3 files at once in a sequence, you need to merge the files from within that playlist in the playlist option itself. In advanced you mark merge files as single file and place them in sequence. this only works if you don't use azuracast managed mode.
That way when the playlist is entered it will play the 3 files in sequence as if it were a single file.

**Comment by hottweelz** [comment]
Well guys, it sucks but we have to move away from Azuracast, loved all the features, but our new Ad provider pays us for triggering ads with 3 sequential injection mp3 files (unlike Targetspot who uses one, but they don't pay anyone ever).

Have to move over to a system we found working... for the sake of income.

Please, PLEASE fix these bugs soon as Azuracast is superior in a lot of ways.

**Comment by SlvrEagle23** [comment]
@hottweelz It doesn't do us much favor to tell us you've left us and to fix the bugs faster, because we're working exactly as hard as we can given our very small team and limited resources. I've personally looked at our code for sequential playlists repeatedly as this issue has been live, and I have continually found _nothing_ wrong with the code, so what's happening is something that is beyond my immediate understanding and thus requires the assistance of another developer.

Unfortunately, this scenario (much like our other AutoDJ scheduling-related problems) is one where there are numerous people springing up to report it as an issue and nobody volunteering to help us actually fix it.

While it bothers me greatly to see people leaving AzuraCast over this, it bothers me even more that this project, made open-source with the explicit goal that "more eyes = better code", has yet to see its open-source nature actually benefit it in cases like this bug. In the immediate term, there is nothing we, as maintainers, can do to create new maintainers.

**Comment by SlvrEagle23** [comment]
@rquara We're testing this out locally and I'm unable to reproduce this as a problem.

As long as the playlist items are ordered correctly on the "Reorder" popup, which controls the actual order of how we play back the sequential items, then our AutoDJ plays them back correctly in that same order.

I noticed from the initial issue that you had the files named 01.mp3, 02.mp3, etc...so I wanted to make sure that you've actually made sure the order is correct on the "Reorder" screen, as that's the order we use when scheduling.

**Comment by Marcioghiraldelli** [comment]
@SlvrEagle23  The sorting problem only happens when you use azuracast managed mode. When you turn off the managed mode by azuracast and let it run by liquidsoap it works as normal. I tested it several times trying to find where the error was and realized that the managed mode, even you ordering, it seems that when azuracast remounts the playlist it messes up the orders. It doesn't seem to be a problem with liquidsoap, but some bug when building the playlist. If you make several lists and put them all in order, you can see in the queue for the next songs that he often doesn't follow the sequence. So you go to the playlist and check the playlist is incorrect. I just managed to reduce this error by putting in the linux cron a method of restarting the radio every day and turning off the mode managed by azuracast.

I wish I could help with programming techniques but my knowledge of this is small. So I try to help with platform testing and verification.

**Comment by SlvrEagle23** [comment]
@Marcioghiraldelli So you're saying when the queue gets reset, _that's_ what is resetting the order of the sequential playlists?

And the order is actually messed up when you go to look at the playlist's order via the "Reorder Playlist" page?

**Comment by Marcioghiraldelli** [comment]
@SlvrEagle23 
I use azuracast a lot with my radios, and I can assure you that one of the playlist Bugs would be this one. You assemble the playlist, organize the sequential part putting in the sequence you want and schedule. In some cases, it doesn't start from the first file, whenever it's mounted for the first time it has this failure of not mounting in the sequence you defined. Then you go to playlist, reorganize and save. The second time he follows, but the third time he messes up the list again. I believe this error happens every time the list is assembled near start.

img1: https://prnt.sc/19a6xks
img2: https://prnt.sc/19a6ziz

**Comment by Marcioghiraldelli** [comment]
In unmanaged mode by azuracast, it presents the error at some point, but it is much more difficult for that to happen. I can solve it by restarting via cron, don't tell me how this solves it because I don't know, I just realized that when it started to give an error of not following, I restarted the radio and went back to normal, so I put in cron the command to restart at dawn to avoid that I keep restarting manually. That way it solved my problem of sequential, but in the mode managed by azuracast it only works well if it's random, sequential when it assembles the list doesn't work.

**Comment by SlvrEagle23** [comment]
@Marcioghiraldelli If it tends to happen only after actually reordering the playlist items, we could indeed reset the "is_queued" status any time a reorder happens...that'll force the system to start over from the top of the playlist in the order you specify.

If there's a separate issue where the actual playlist order (as seen on the Reorder page) goes out of order at some point, that's something we'd certainly need to investigate and is likely the cause of this issue.

**Comment by Marcioghiraldelli** [comment]
So I notice that in the playlist view it changes the order. EVEN if you put 01,02,03 , when it assembles the list it changes putting 02,01,03 or 03,01,02 , something like that. Putting yourself at the beginning of the playlist whenever it is assembled can solve some problems, but this part of changing the sequence when you assemble, even if you configure the sequence, I don't think it would. The strange thing is that, it's always the first time it assembles the list, after it happens, you'll check and it's all out of sequence, so you fix it again and the second time it even follows but the third time it already changes the sequence.
As I said I really wanted to help in the programming part but I don't have the knowledge for that.

**Comment by SlvrEagle23** [comment]
@Marcioghiraldelli I mostly understand what you mean. I'm looking into it locally. I haven't found this to be the case in my local tests yet, but I will update if I discover anything different.

**Comment by Marcioghiraldelli** [comment]
@SlvrEagle23  Perfect!
I see this problem when I create multiple playlists and they are all sequential.

**Comment by hottweelz** [comment]
@SlvrEagle23 I'm sure its frustrating when you can't get the help you need, but at the same time, Azuracast is offered by Digital Ocean as marketplace VM.

You can't possibly expect all of your users to be developers in the language and coding that you use.

If you want to replicate the issue that we are all suffering from... stop using your local host system.

Go to Digital Ocean, install an Azuracast VM and test it out... you'll replicate it... (I've deployed at least 7 on my own)

Apologies for the leaving statement but wanted to let you know that we DO NOT WANT to leave... but $ talks.  Azuracast is GREAT for many other reasons.

**Comment by SC2Mitch** [comment]
The problem isn't the 'local host system', I can assure you, I have spent the better part of the weekend trying to debug and investigate this issue on my own production server, which is hosted by Digital Ocean and OVH respectively, alas these errors or sequential issues simply did not appear. 

And to clarify one thing, Digital Ocean doesn't pay AzuraCast to be on the market *to my knowledge*, I'm fairly sure all developers can apply to be on it.

**Comment by Marcioghiraldelli** [comment]
@SC2Mitch  I can't understand how you and @SlvrEagle23  can't replicate the error if most people using azuracast experience it. There would be no reason why we are here trying to come up with a solution if the problem really didn't exist.
It's easy to get into trouble, schedule at least 10 playlists and put them all in sequentially. and leave a general rotation.
In these 10 playlist renames the files to 1,2,3,4;.
Place all of them sequentially and schedule.
The error occurs when you schedule a playlist and set it to sequential. I noticed that when azuracast manages the auto dj when it goes to assemble the list, it ends up not putting in the right sequence that you previously configured playing out of sequence.

**Comment by SC2Mitch** [comment]
We have tried this, I can assure you, this is from my production instance, right now. 
![image](https://user-images.githubusercontent.com/29955942/124831554-4d892a00-df73-11eb-9ec7-8055e0381a95.png)

We cannot reproduce this, locally.

**Comment by Marcioghiraldelli** [comment]
in the test you are doing there are few songs within a playlist, each playlist must have at least 5 songs. And having different playlists, that is, each playlist with 5 different files.
The way you're trying to replicate is basic, maybe that's why you can't get to the error. I say this because I also have radio in production and I have 23 playlists. Usually my calendar lists where the error happens has 5 files that I put in sequence.

**Comment by hottweelz** [comment]
Would it help if we gave you access to one of ours?

**Comment by Vaalyn** [comment]
Just wanted to chime to provide some further explanation of what we need in order to better reproduce issues like this / issues with the AutoDJ / Scheduler.

Some things have been answered in this thread already but I wanted to provide our users with a small checklist that can be followed since a lot of the time we only get reports with minimal information in regards to how to reproduce it. Then when we are unable to reproduce the issue with the limited information we got we need to follow up with the user/s reporting it in order to try getting more information about their setup. This is often times flawed since some things that are seemingly obvious to the user/s are not obvious to us since we can't see their system and haven't worked with their setup so we aren't familiar with what they are working with.

Most of the time we would need the following information to have a high chance of being able to easily reproduce a problem:

- Amount of active playlists you are using when this happens
- Exact configuration details of those playlists
- Amount of media files in those playlists
- Rough average length of the media files
- Are there any unusual files mixed in there
    - Example: most files are ~4-5m long but some are x hours
- Are you using auto-assign folders
- Station AutoDJ configuration (if anything is not on the default values)
- How long / how many playthroughs of a playlist does it need to manifest the issue
- Are song requests happening in the time the issue shows
- How exactly does the playlists internal "Playback Queue" look and change while the issue happens

I'll also take a shot at reproducing this problem later today so if someone of you can provide me with a description of an AzuraCast station that should exhibit this issue I'd greatly appreciate it.

(Just tell me "create x playlists of type x with config x, put x media files of x length into playlist a, y in b, z in c, etc... then do this, this and this, wait for x minutes and you should see that xyz happens).

**Comment by Marcioghiraldelli** [comment]
Usually the problem happens when you have around 10 to 15 playlists.
All in sequence.
It can be with names of 1,2,3,4 or 01,02,03,04.
The problem happens in both. Playlists are around 20 to 30 minutes long, usually with quantities of 4 files or more.

On the radios where I did the test, there were 23 playlists.
All with an average total time of 30 minutes to 1 hour.
I noticed that the problem happens whenever you organize the playlist sequence and schedule it.
It only happens on scheduled playlists.
When it's about to enter the timetable, azuracast assembles the playlist but completely out of the order you organized it.
So you arrange it again and the second time you log in, it usually logs in right but the third time the problem happens again.
Often it even takes files from general rotation or from other playlists.

**Comment by Marcioghiraldelli** [comment]
once again I say.
The problem exists and many people face this problem.

**Comment by sidney65** [comment]
One important thing to know, is if you make a playlist whith 1 track, is work fine.
If you make a playlist with more tracks (3 or more for example), only the first one are play.

**Comment by Marcioghiraldelli** [comment]
@sidney65  in mine it even plays all the songs on the list but not in the sequence that I defined but randomly ie I organize the list but azuracast redoes the list wrongly.

**Comment by SlvrEagle23** [comment]
> You can't possibly expect all of your users to be developers in the language and coding that you use.

We certainly don't, and we build AzuraCast to be usable for station operators who aren't technically skilled in the same languages that we use to build the application.

All the same, it would be nice if _any_ of our users were developers who were willing and able to contribute to our project the way our very small team does. It isn't at all unreasonable for us to say that having more help and more hands on our project would make it a better one for everyone, especially with these elusive, hard-to-pin-down problems we have with things like scheduling and playlist rotation.

> once again I say.
> The problem exists and many people face this problem.

We aren't denying the existence of the problem, but every time we dig into the code and try to reproduce it, we come up short one way or another. My local instance, for example, schedules everything _perfectly_. It does not now, and hasn't for months or years now, had a scheduling problem. I wish I could just give you all copies of my local installation to use (that's actually what our Docker images are intending to do, but alas they aren't _identical_).

Believe me, we're frustrated with the trouble this is causing users, too.

But this isn't a reason to give up on AzuraCast or otherwise write it off as poorly maintained software. That isn't a fair characterization of the problem. The problem isn't caused by our incompetence or anything like that, we're just running into a very familiar problem where some bugs are extremely elusive, hard to pin down, and never seem to happen when observed by developers who can fix them.

**Comment by SlvrEagle23** [comment]
> Would it help if we gave you access to one of ours?

@hottweelz If you have a station that's demonstrating the scheduling problem with a sequential playlist, and it's reliably demonstrating it right now...then absolutely yes, DM me on Discord and I'll take a personal look at your installation.

**Comment by hottweelz** [comment]
> > Would it help if we gave you access to one of ours?
> 
> @hottweelz If you have a station that's demonstrating the scheduling problem with a sequential playlist, and it's reliably demonstrating it right now...then absolutely yes, DM me on Discord and I'll take a personal look at your installation.

I messaged you on Twitter as I can't seem to find your Discord.

"Hey, I cant seem to find your Discord, I'm "hottweelz" that had the issue with "sequential playlists" because I really did honestly have to abandon Azuracast, I lost track of the github issue.  I MISS Azuracast and would love to see it fixed... I offered you access to replicate the issue if you needed it.  Still want it?"

**Comment by SlvrEagle23** [comment]
@hottweelz I've gone back and tested this issue locally with the latest Rolling Release version of AzuraCast, and it appears to work perfectly, as in, any playback interval (once per x songs, minutes, etc.) will still follow the order of a sequential playlist.

Whatever bug was causing this issue may have been fixed in recent updates that we've made to the scheduler. I'd recommend updating to the latest Rolling Release version and seeing if the issue is solved, as it very well may be for your scenario as well.

**Comment by wmfinck** [comment]
I am brand new to Azuracast, and want it to replace a few streams which I have always managed manually. That is mainly because the liquidsoap 2.x upgrade gives many errors for my old version 1.x scripts. 

So I made 7 sequential playlists, one for each day of the week, and carefully ordered the recordings, as it is important to me that they play in a prescribed order. I added each to the playlist in the precise order I want them to play. Each list is very close to 12 hours of play and contains 12-15 items. Some are songs, some are 2-hour podcast recordings. 

Immediately, Azuracast began changing the order of the playlists. Moving all songs to the end, randomly changing the order of long recordings. The songs are in a different folder. One thing I noticed is that when a recording is added to a playlist from the media page, every playlist which contains that recording is reordered. 

I just now used the reorder feature to put all 7 playlists back into the correct order. I am hoping they stay in that order, or I will never be able to migrate my streams away from the old system. But if I add one recording to a different, new playlist, its order changes on EVERY playlist in which it appears! Usually it is moved to the end of every playlist, even if it was at the beginning. WTF?

**Comment by BusterNeece** [comment]
@wmfinck This isn't an invalid bug report on your part...I can definitely see how this could happen, especially since, under the hood, our "set playlists" functionality clears all playlists from media and then sets the playlists on all of the selected media.

However, it would certainly be more useful for us to track if you made a new issue, as this is a unique (user-experience-based) issue rather than the scheduler actually malfunctioning.

**Comment by wmfinck** [comment]
I have just created a new issue. Thank you!

**Comment by pavlos2** [comment]
Hello everyone,

I would like to ask if the "sequential playlist" issue can be addressed and how? I made multiple attempts to resolve it, including reinstalling Azuracast on a new Linode server. My setup involves numerous playlists, with one of them containing over 1,000 files that need to be played in sequential mode. Despite using the Reorder function and confirming that the files are listed in the correct order, they still play in a completely different order. I have tried resetting the cache, queues, and even reinstalling Azuracast, but the problem persists.

I would appreciate any guidance or solution on this matter.

Thank you in advance!
